Pg datatype varchar

faviansamatha opened the pull request:
### Description

[Describe what this change achieves]
 
### Issues Resolved

[List any issues this PR will resolve]


### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

faviansamatha committed:
Added initial psqlodbc datatype test for bigint

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Created table in master_dbo schema

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Refactored the tests, added more test cases, and split them into multiple tests

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Initial commit for varchar tests

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Addressed PR comments

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Added arithmetic and view creation tests

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
Added initial varchar tests. Still need to clean up

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
added table_creat_fail test

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha committed:
deleted bigint

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

faviansamatha commented on test/odbc/test_pg_datatype_varchar.cpp:
```diff
@@ -0,0 +1,668 @@
+  "oNNiA5koeFYidyobnjFDKvyJ4XNfQRIUIfJfa8YGCfzmw35UYisdkyhfsHJGCPiC9bhBke4oTVmg9cvRfAL3xdMfaxA55f392Vbupz0wgSPfhrg6IUOBmAqR65zYeDfcM"
+  "plb4gM60zoHm2Oe1ohrvO6xkDX9rOrrvCSpVInbcYE3iyzTyG97bVp19ywZLq7DmyT9sbVLjX8tXtl6KFmDngAIs9xOxjpbkPRWmbOnvDEhx9xdK6WCUUXYUybmCj0fgI"
+  "BLzq2nAqMA8tjpKHxG1qp95Ii14hiDFkSHjMa7X5K4Y9tu5FPLsucvo6wKLFqdwETdDVmzeACxCQCRt7EtoPQGaBVT5z3jKniGJKWQFyqQmC3hBsisudBum6myRXwz2eI"
+  "Fzq4QR9PKRVERBAwEA9BDkpiCpjbG2uavD9oFvtF9LGFErLu6XmiMTcP1S5";
```
Not sure how to make this string look prettier. Could make a deterministic function that generates a string of n characters. Any thoughts on that?

rossd-bitquill replied:
We're going to need to do this a few more times (nvarchar, at least, probably a few others).  I think it would be worth writing a short common routine that will take a length as an argument and spit out a string of the desired length.

As you mention, for testing purposes it should be deterministic, but we could also just have the seed passed in as an additional argument.

faviansamatha committed:
small cleanup

Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

jusjosj committed:
Expand explain for PRINT, THROW, USE and RAISERROR (#561)

This commit extends the functionality of babelfish_showplan_all to include print, throw, raiserror, and use
This is largely done in a manner consistent with the other explain functions existing throughout the codebase.

Task: BABEL-2902
Signed-off-by: Justin Jossick <jusjosj@amazon.com>

shah-nirmit committed:
Change github actions to point to correct engine dev branch (#649)

Authored-by: Nirmit Shah <nirmisha@amazon.com>
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>

Sairakan committed:
Fix table hint parsing (#637)

Previously, Babelfish would sometimes incorrectly parse a query that used a table alias with column definitions as a table hint and raise an error. This commit fixes the issue by having the parser only accept a specific defined list of keywords as possible table hints.

N.B. Babelfish will report incorrect table hint syntax as generic syntax errors instead of specifically table hint errors as SQL Server does. This commit does not attempt to mimic this behavior, as that would require a more extensive modification to the antlr parser (most likely requiring a new grammar rule to be added).

Task: BABEL-2805
Signed-off-by: Jason Teng jasonten@amazon.com

AmithKamath committed:
Add JDBC tests for CONVERT()

An issue BABEL-2054 was reported where CONVERT(datetime, string) failed
intermittently. However, this issue is not reproducible and this commit
adds JDBC tests to validate the scenario mentioned in the JIRA

Task: BABEL-2054

Signed-off-by: Amith Kamath <amithkam@amazon.com>

wboettge committed:
Login who owns database should always be dbo also without sysadmin role

In SQL Server, the login that creates a database is automatically the
database owner. This means that the login can always access the
databases they create.

However, previously in babelfish, if a login is stripped of the sysadmin
role after creating a database, and the login does not have a
user associated with the database, then the login can no longer access
the database. Now, when a login attempts to access and there
is not an associated user, the login will be considered the dbo and
allowed access.

Task: BABEL-3548

Signed-off-by: Walt Boettge <wboettge@amazon.com>

shah-nirmit committed:
Refactor Encoding Conversion Related Functions to improve Performance (#632)

Previously, we were using pg_any_to_server() API on receiving side for varchar, char and text datatypes.
pg_any_to_server() internally do the sys cache lookup which was expensive in some cases(varchar,char and text datatypes) and we really do not need to do it as TDS only supports collations which requires only few set of encodings. This commit adds new function encoding_conv_util() which would take src_encoding and dest_encoding as argument and will call appropriate encoding conversion functionality based on those arguments. This commit also copies few encoding conversion related functionality (TsqlLocalToUtf)from engine to common extension and extends them to return byte length of encoded string. 
 
Changes: 
1.  Remove duplicate encoding logic from babelfishpg_tsql extension
2. Introduce conditional handling for encoding conversion for Varchar, Char and Text datatypes on receiver side
3. Copy and enhance TsqlLocalToUtf from engine side to be used by various encoding functions


### Testing
- Collation TestFramework
  - Tested Collations
    - sql_latin1_general_cp1_cs_as
    - arabic_cs_as
    - greek_cs_as
    - turkish_cs_as
    - polish_cs_as
    - vietnamese_cs_as
    - ukrainian_cs_as
    - japanese_cs_as
    - korean_wansung_cs_as
- Manual testing with custom C# program

Task: BABEL-3542
Signed-off-by: Nirmit Shah <nirmisha@amazon.com>

ZhenyeLee committed:
Fix syscoumns test failure (#660)

The helper function in prepare file searching object_name by oid in both pg_class and pg_proc. However, pg_class and pg_proc may have the same oid and it cause test failure. In this PR, we break down the search into to functions, one is in pg_class, and another one is in pg_proc. And also removed order by id from verification file, because oid can be changed after MVU.

Signed-off-by: Zhenye Li zhenye@amazon.com

Commit:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into BABEL_2_X_DEV

Commit:
rebase

Commit:
Comparison operators added, String operators, unique and composite key added
